### DISCUSSION TITLE
Managing configuration and Github best practices across all repos?

### DISCUSSION THREAD

**Discussion by kgpayne** [MeltanoLabs/Meta#26] [General]
My questions are:

1. Would we benefit from some kind of centralisation of config for the repos in MeltanoLabs (beyond the existing 'Teams')?
2. If so, what tools are available to facilitate centralised configuration management of multiple Github repos?

Background:

As I have been working on adding poetry-dynamic-versioning to 2 or 3 repos in MeltanoLabs, I have needed to request `Admin` level privileges in order to add PyPI secrets. I have also noticed differences in features enabled (e.g. use of branch protection), even though the use cases don't vary. As we now have have 30 repos under this Org, I believe it would be helpful to look at options to centralise and standardise management of our GH repo settings (including team/user access, protected branches, enabled features such as Wiki etc.).

A first pass led me to the [github terraform provider](https://github.com/integrations/terraform-provider-github), but there may be other Github-native tool options too.

cc @MeltanoLabs/engineering

**Comment by aaronsteers** [comment]
> A first pass led me to the [github terraform provider](https://github.com/integrations/terraform-provider-github), but there may be other Github-native tool options too.

I love this idea! A quick google search surfaced this similar [GitHub provider for Pulumi](https://www.pulumi.com/registry/packages/github/api-docs/repository/). Since we all know Python, and since we already are cross-training the full team on Pulumi, the Pulumi/Python pattern would be my preference unless there are siginificant gaps in the Pulumi-based provider.

Using the IAC approach, I imagine we could setup a workflow in [internal-tech-ops](https://github.com/meltano/internal-tech-ops) which operates on all repos we list from `meltano` and `meltanolabs` repos.

1. Since the repos already exist, we'd need to first `import` or `get` resources and then apply changes.
2. We'd declare our preferred settings for each repo, including common settings like `allow_merge_commit`, `allow_auto_merge`, and other configuration details.
3. We could define groups of differently-configured repos as needed; not all repos would need to have identical configurations - but likely we could manage all repos with 2-4 total configsets.
